be/src/connector/mysql_connector.cpp

```diff
@@ -0,0 +1,463 @@
+    bool mysql_eos = false;
+    int row_num = 0;
+
+    while (true) {
+        RETURN_IF_CANCELLED(state);
+
+        if (row_num >= state->chunk_size()) {
+            return Status::OK();
+        }
+
+        // read mysql
+        char** data = nullptr;
+        size_t* length = nullptr;
+        RETURN_IF_ERROR(_mysql_scanner->get_next_row(&data, &length, &mysql_eos));
+        if (mysql_eos) {
+            _is_finished = true;
+            return Status::OK();
+        }
+
+        ++row_num;
+        RETURN_IF_ERROR(fill_chunk(chunk, data, length));
+        ++_rows_read;
+    }
+}
+
+int64_t MySQLDataSource::raw_rows_read() const {
+    return _rows_read;
+}
+
+int64_t MySQLDataSource::num_rows_read() const {
+    return _rows_read;
+}
+
+void MySQLDataSource::close(RuntimeState* state) {
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+}
+
+Status MySQLDataSource::fill_chunk(vectorized::ChunkPtr* chunk, char** data, size_t* length) {
+    int materialized_col_idx = -1;
+    for (size_t col_idx = 0; col_idx < _slot_num; ++col_idx) {
+        SlotDescriptor* slot_desc = _tuple_desc->slots()[col_idx];
+        ColumnPtr column = (*chunk)->get_column_by_slot_id(slot_desc->id());
+
+        // because the fe planner filter the non_materialize column
+        if (!slot_desc->is_materialized()) {
+            continue;
+        }
+
+        ++materialized_col_idx;
+
+        if (data[materialized_col_idx] == nullptr) {
+            if (slot_desc->is_nullable()) {
+                column->append_nulls(1);
+            } else {
+                std::stringstream ss;
+                ss << "nonnull column contains NULL. table=" << _table_name << ", column=" << slot_desc->col_name();
+                return Status::InternalError(ss.str());
+            }
+        } else {
+            RETURN_IF_ERROR(append_text_to_column(data[materialized_col_idx], length[materialized_col_idx], slot_desc,
+                                                  column.get()));
+        }
+    }
+    return Status::OK();
+}
+
+Status MySQLDataSource::append_text_to_column(const char* data, const int& len, const SlotDescriptor* slot_desc,
+                                              Column* column) {
+    // only \N will be treated as NULL
+    if (slot_desc->is_nullable()) {
+        if (len == 2 && data[0] == '\\' && data[1] == 'N') {
+            column->append_nulls(1);
+            return Status::OK();
+        }
+    }
+
+    Column* data_column = column;
+    if (data_column->is_nullable()) {
+        auto* nullable_column = down_cast<NullableColumn*>(data_column);
+        data_column = nullable_column->data_column().get();
+    }
+
+    bool parse_success = true;
+
+    // Parse the raw-text data. Translate the text string to internal format.
+    switch (slot_desc->type().type) {
+    case TYPE_VARCHAR:
+    case TYPE_CHAR: {
+        Slice value(data, len);
+        dynamic_cast<BinaryColumn*>(data_column)->append(value);
```

Comment:
dynamic_cast will cause bad performance if you call it for each column. Maybe you can use represent_cast